File: android/app/src/main/java/edu/gatech/cs2340/rajsjarfiles/spacetrader/entity/market/Item.java
package edu.gatech.cs2340.rajsjarfiles.spacetrader.entity.market;

public class Item {

    private Good good; //Types of good
    private int quantity; //Quantity of good
    private int price; //Price of good (price when traded)

    Item(ItemBuilder builder) {
        good = builder.good;
        quantity = builder.quantity;
        price = builder.price;
    }

    public String getGoodName() {
        return good.getName();
    }

    public Good getGood() {
        return this.good;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public int getPrice() {
        return price;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }
    public void addQuantity(int quantity) {
        this.quantity += quantity;
    }
    public void subQuantity(int quantity) {this.quantity -= quantity; }

    @Override
    public String toString() {
        return quantity + " " + good.getName() + " at " + price + " units each.";
    }

    public static class ItemBuilder{
        private Good good;
        private int quantity = 1;
        private int price = 0;

       public ItemBuilder(Good tg) {
            this.good = tg;
       }

       public ItemBuilder quantity(int quantity) {
           this.quantity = quantity;
           return this;
       }

       public ItemBuilder price(int price) {
           this.price = price;
           return this;
       }

       public Item build() {
            return new Item(this);
        }
    }
}

File: ios/SpaceTrader/SpaceTrader/src/entity/market/Item.swift
//
//  .swift
//  SpaceTrader
//
//  Created by Thomas Suarez on 4/20/19.
//  Copyright © 2019 RajsJarFiles. All rights reserved.
//

import Foundation

class Item : CustomStringConvertible {
    
    var good : Good
    var quantity : Int
    var price : Int
    
    init(builder: ItemBuilder) {
        self.good = builder.good
        self.quantity = builder.quantity
        self.price = builder.price
    }
    
    // Java-like Getters and Setters
    func getGoodName() -> String {
        return self.good.getName()
    }
    func addQuantity(quantity : Int) {
        self.quantity += quantity
    }
    func subQuantity(quantity : Int) {
        self.quantity -= quantity
    }
    
    var description: String {
        return self.quantity + " " + self.good.getName() + " at " + self.price + " units each."
    }
    
    class ItemBuilder {
        
        var good : Good
        var quantity : Int = -1
        var price : Int = 0
        
        init(tg : Good) {
            self.good = tg
        }
        
        func quantity(_ quantity : Int) -> ItemBuilder {
            self.quantity = quantity
            return self
        }
        
        func price(_ price : Int) -> ItemBuilder {
            self.price = price
            return self
        }
        
        func build() -> Item {
            return Item(builder: self)
        }
        
    }
    
}
